laplace/baselaplace.py

```diff
@@ -1792,8 +1881,1041 @@ def sample(
 
 
 class FunctionalLaplace(BaseLaplace):
-    pass
+    """Applying the GGN (Generalized Gauss-Newton) approximation for the Hessian in the Laplace approximation of the posterior
+    turns the underlying probabilistic model from a BNN into a GLM (generalized linear model).
+    This GLM (in the weight space) is equivalent to a GP (in the function space), see
+    [Approximate Inference Turns Deep Networks into Gaussian Processes (Khan et al., 2019)](https://arxiv.org/abs/1906.01930)
+
+    This class implements the (approximate) GP inference through which
+    we obtain the desired quantities (posterior predictive, marginal log-likelihood).
+    See [Improving predictions of Bayesian neural nets via local linearization (Immer et al., 2021)](https://arxiv.org/abs/2008.08400)
+    for more details.
+
+    Note that for `likelihood='classification'`, we approximate \( L_{NN} \\) with a diagonal matrix
+    ( \\( L_{NN} \\) is a block-diagonal matrix, where blocks represent Hessians of per-data-point log-likelihood w.r.t.
+    neural network output \\( f \\), See Appendix [A.2.1](https://arxiv.org/abs/2008.08400) for exact definition). We
+    resort to such an approximation because of the (possible) errors found in Laplace approximation for
+    multiclass GP classification in Chapter 3.5 of [R&W 2006 GP book](http://www.gaussianprocess.org/gpml/),
+    see the question
+    [here](https://stats.stackexchange.com/questions/555183/gaussian-processes-multi-class-laplace-approximation)
+    for more details. Alternatively, one could also resort to *one-vs-one* or *one-vs-rest* implementations
+    for multiclass classification, however, that is not (yet) supported here.
+
+    Parameters
+    ----------
+    num_data : int
+        number of data points for Subset-of-Data (SOD) approximate GP inference.
+    diagonal_kernel : bool
```

Comment:
I feel like `independent_kernel` or `independent_outputs` would be more suitable variable name. `diagonal_kernel` sounded to me like the extreme case a subset of a single data point so no correlations are estimated. What do you think?

Reply:
It is true the name might be confusing, I prefer `independent_outputs` rather than `independent_kernel`  as in the latter it might be unclear what "independent" means.